qiskit_addon_utils/exp_vals/expectation_values.py

```diff
@@ -0,0 +1,472 @@
+# This code is a Qiskit project.
+#
+# (C) Copyright IBM 2025.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Utility functions for calculating expectation values of observables."""
+
+from __future__ import annotations
+
+import numpy as np
+from qiskit.primitives import BitArray
+from qiskit.quantum_info import Pauli, PauliLindbladMap, SparseObservable, SparsePauliOp
+
+
+def expectation_values(
+    bool_array: np.ndarray[tuple[int, ...], np.dtype[np.bool]],
+    basis_dict: dict[Pauli, list[SparsePauliOp | None]],
+    meas_basis_axis: int | None = None,
+    avg_axis: int | tuple[int, ...] | None = None,
+    meas_flips: np.ndarray[tuple[int, ...], np.dtype[np.bool]] | None = None,
+    pec_signs: np.ndarray[tuple[int, ...], np.dtype[np.bool]] | None = None,
+    postselect_mask: np.ndarray[tuple[int, ...], np.dtype[np.bool]] | None = None,
+    pec_gamma: float | None = None,
+    rescale_factors: list[list[float]] | None = None,
```

Comment:
`rescale_factors_combined = [np.array([obs for base_obs in base for obs in base_obs]) for base in rescale_factors]`

The line above (currently 176) suggests this should be a 3D list of floats. @aeddins-ibm , could you please confirm?